lib/srv/discovery/discovery.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/cloud"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/server"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+type Config struct {
+	Clients     cloud.Clients
+	Matchers    []services.AWSMatcher
+	NodeWatcher *services.NodeWatcher
+	Emitter     apievents.Emitter
+	AccessPoint auth.DiscoveryAccessPoint
+}
+
+// Server is a discovery server, used to discover cloud resources for
+// inclusion in Teleport
+type Server struct {
+	*Config
+	ctx context.Context
+
+	// log is used for logging.
+	log *logrus.Entry
+
+	// cloudWatcher periodically retrieves cloud resources, currently
+	// only EC2
+	cloudWatcher *server.Watcher
+	// ec2Installer is used to start the installation process on discovered EC2 nodes
+	ec2Installer *server.SSMInstaller
+}
+
+func New(ctx context.Context, cfg *Config) (*Server, error) {
+	s := &Server{
+		Config: cfg,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component:       teleport.ComponentDiscovery,
+			trace.ComponentFields: logrus.Fields{},
+		}),
+		ctx: ctx,
```

Comment:
I would wrap the parent context in a local context with a cancel function so discovery service can be stopped independently of the outer process.